packages/loopring_v3/contracts/amm/libamm/AmmJoinRequest.sol

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/MathUint96.sol";
+import "../../thirdparty/SafeCast.sol";
+import "./AmmData.sol";
+import "./AmmUtil.sol";
+
+
+/// @title AmmJoinRequest
+library AmmJoinRequest
+{
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    bytes32 constant public POOLJOIN_TYPEHASH = keccak256(
+        "PoolJoin(address owner,bool fromLayer2,uint256 minPoolAmountOut,uint256[] maxAmountsIn,uint32[] storageIDs,uint256 validUntil)"
```

Comment:
Probably still TODO, but the fee not part of signed message yet.

Reply:
fixed